### PR TITLE
Allow picking a custom background.

### PR DESCRIPTION
**Pull request by xfix**

### CODE REVIEW

File: js/client.js [hunk 1]
```diff
@@ -2686,4 +2690,34 @@
...
+			buf += '<input type="file" accept="image/*" name="bgfile">';
+
+			buf += '<p><button name="close">Cancel</button></p>';
+			this.$el.html(buf);
+		},
+		setBg: function(e) {
+			var file = e.currentTarget.files[0];
+			var reader = new FileReader();
+			reader.onload = function (e) {
+				e.target.result;
```
Comment (kotarou3) [inline js/client.js]: ?

**Comment by kotarou3** [comment]
Also, you might want allow for URI entry as well

File: js/client.js [hunk 2]
```diff
@@ -2686,4 +2690,34 @@
...
+				var bg = '#344b6c url(' + e.target.result + ') no-repeat left center fixed';
+				Tools.prefs('bg', bg);
```
Comment (kotarou3) [inline js/client.js]: Be careful of running out of `localStorage` space

File: js/client.js [hunk 3]
```diff
@@ -2686,4 +2690,33 @@
...
+
+			buf += '<p><button name="close">Cancel</button></p>';
+			this.$el.html(buf);
+		},
+		setBg: function(e) {
+			var file = e.currentTarget.files[0];
+			var reader = new FileReader();
+			reader.onload = function (e) {
+				var bg = '#344b6c url(' + e.target.result + ') no-repeat left center fixed';
+				Tools.prefs('bg', bg);
```
Comment (kotarou3) [inline js/client.js]: Be careful of running out of `localStorage` space

**Comment by Slayer95** [comment]
Cool!

**Comment by xfix** [comment]
@kotarou3: URI entry can be added separately if someone is interested in doing that. Currently I'm implementing just the basic feature.

(also, web browser file pickers should allow URLs, but currently the only browser I know that allows that is Konqueror... that's outside of scope of this pull request however)

File: js/client.js [hunk 4]
```diff
@@ -2686,4 +2690,42 @@
...
+			buf += '<p><button name="close">Cancel</button></p>';
+			this.$el.html(buf);
+		},
+		setBg: function(e) {
+			$('.bgstatus').text('Changing background image.');
+			var file = e.currentTarget.files[0];
+			var reader = new FileReader();
+			var self = this;
+			reader.onload = function(e) {
+				var bg = '#344b6c url(' + e.target.result + ') no-repeat left center fixed';
```
Comment (Zarel) [inline js/client.js]: Data URL is actually a pretty interesting idea.

**Comment by Zarel** [comment]
My main fear is that this is storing too much data in localStorage and/or RAM

**Comment by xfix** [comment]
Hm, that's understandable. RAM usage shouldn't be too much of an issue, an image has to be stored in RAM anyway to display it. I guess the image is stored essentially twice, but with how much RAM is in today's machines, it shouldn't be an issue (the image is 3.5MB max).

As for localStorage limit, it's 5MB, except for Internet Explorer 8 which has 128 KB limit. I guess this could be using IndexedDB instead of localStorage, because IndexedDB has higher size limits (50MB), however.

**Comment by Zarel** [comment]
`localStorage` limit may be 5MB, but larger data amounts tend to get pruned more often afaik.

This only affects people who actually use custom backgrounds, though, so I don't think it's too much of a problem.

**Comment by Zarel** [comment]
You might want to warn about large images, though? `base64` is like 75% efficient, and we use `localStorage` for other things, so we might want to warn at 1MB or 1MB or something.

**Comment by Zarel** [comment]
(I'd recommend making a new pull request for that)

**Pull request merged**